core/remote_state_verifier.go

```diff
@@ -0,0 +1,399 @@
+	if vm.verifiedCache.Len() >= verifiedCacheSize {
+		vm.verifiedCache.RemoveOldest()
+	}
+	vm.verifiedCache.Add(hash, true)
+}
+
+// AncestorVerified function check block has been verified or it's a empty block.
+func (vm *remoteVerifyManager) AncestorVerified(header *types.Header) bool {
+	// find header of H-11 block.
+	header = vm.bc.GetHeaderByNumber(header.Number.Uint64() - maxForkHeight)
+	// If start from genesis block, there has not a H-11 block.
+	if header == nil {
+		return true
+	}
+	// check whether H-11 block is a empty block.
+	if header.TxHash == types.EmptyRootHash {
+		parent := vm.bc.GetHeaderByHash(header.ParentHash)
+		return header.Root == parent.Root
+	}
+	hash := header.Hash()
+	_, exist := vm.verifiedCache.Get(hash)
+	return exist
+}
+
+func (vm *remoteVerifyManager) HandleRootResponse(vr *VerifyResult, pid string) error {
+	vm.messageCh <- verifyMessage{verifyResult: vr, peerId: pid}
+	return nil
+}
+
+type VerifyResult struct {
+	Status      types.VerifyStatus
+	BlockNumber uint64
+	BlockHash   common.Hash
+	Root        common.Hash
+}
+
+type verifyMessage struct {
+	verifyResult *VerifyResult
+	peerId       string
+}
+
+type verifyTask struct {
+	diffhash       common.Hash
+	blockHeader    *types.Header
+	candidatePeers verifyPeers
+	BadPeers       map[string]struct{}
+	startAt        time.Time
+	allowUntrusted bool
+
+	messageCh  chan verifyMessage
+	terminalCh chan struct{}
+}
+
+func NewVerifyTask(diffhash common.Hash, header *types.Header, peers verifyPeers, verifyCh chan common.Hash, allowUntrusted bool) *verifyTask {
+	vt := &verifyTask{
+		diffhash:       diffhash,
+		blockHeader:    header,
+		candidatePeers: peers,
+		BadPeers:       make(map[string]struct{}),
+		allowUntrusted: allowUntrusted,
+		messageCh:      make(chan verifyMessage),
+		terminalCh:     make(chan struct{}),
+	}
+	go vt.Start(verifyCh)
+	return vt
+}
+
+func (vt *verifyTask) Start(verifyCh chan common.Hash) {
+	vt.startAt = time.Now()
+
+	vt.sendVerifyRequest(defaultPeerNumber)
+	resend := time.NewTicker(resendInterval)
+	defer resend.Stop()
+	for {
+		select {
+		case msg := <-vt.messageCh:
+			switch msg.verifyResult.Status.Code / 100 {
+			case 1:
+				switch msg.verifyResult.Status {
+				case types.StatusFullVerified:
+					vt.compareRootHashAndWrite(msg, verifyCh)
+					newRecievedMsgTypeGauge("fullVerified", msg.peerId).Inc(1)
+				case types.StatusPartiallyVerified:
+					log.Warn("block %s , num= %s is untrusted verified", msg.verifyResult.BlockHash, msg.verifyResult.BlockNumber)
+					if vt.allowUntrusted {
+						vt.compareRootHashAndWrite(msg, verifyCh)
+					}
+					newRecievedMsgTypeGauge("partialVerified", msg.peerId).Inc(1)
+				}
+
+			case 2, 4:
+				vt.BadPeers[msg.peerId] = struct{}{}
+				log.Info("peer %s is not available: code %d, msg %s,", msg.peerId, msg.verifyResult.Status.Code, msg.verifyResult.Status.Msg)
+				switch msg.verifyResult.Status {
+				case types.StatusDiffHashMismatch:
```

Comment:
Too many duplicated code, please use map to reduce it.

Reply:
fixed